samplePDF/samplePDFFDBase.cpp

```diff
@@ -51,6 +41,168 @@ samplePDFFDBase::~samplePDFFDBase()
+      throw MaCh3Exception(__FILE__, __LINE__);
+    }
+    nDimensions++;
+  }
+  
+  if(nDimensions == 0){
+    MACH3LOG_ERROR("Error setting up the sample binning");
+    MACH3LOG_ERROR("Number of dimensions is {}", nDimensions);
+    MACH3LOG_ERROR("Check that an XVarStr has been given in the sample config");
+    throw MaCh3Exception(__FILE__, __LINE__);
+  } else{
+    MACH3LOG_INFO("Found {} dimensions for sample binning", nDimensions);
+  }
+  
+  //Sanity check that some binning has been specified
+  if(SampleXBins.size() == 0 && SampleYBins.size() == 0){
+    MACH3LOG_ERROR("No binning specified for either X or Y of sample binning, please add some binning to the sample config {}", SampleManager->GetFileName());
+    throw MaCh3Exception(__FILE__, __LINE__);
+  }
+  
+  //FD file info
+  if (!CheckNodeExists(SampleManager->raw(), "InputFiles", "mtupleprefix")){
+    MACH3LOG_ERROR("InputFiles:mtupleprefix not given in {}, please add this", SampleManager->GetFileName());
+  }
+  mtupleprefix = SampleManager->raw()["InputFiles"]["mtupleprefix"].as<std::string>();
+  mtuplesuffix = SampleManager->raw()["InputFiles"]["mtuplesuffix"].as<std::string>();
+  splineprefix = SampleManager->raw()["InputFiles"]["splineprefix"].as<std::string>();
+  splinesuffix = SampleManager->raw()["InputFiles"]["splinesuffix"].as<std::string>();
+  
+  for (auto const &osc_channel : SampleManager->raw()["SubSamples"]) {
+    mtuple_files.push_back(mtupleprefix+osc_channel["mtuplefile"].as<std::string>()+mtuplesuffix);
+    spline_files.push_back(splineprefix+osc_channel["splinefile"].as<std::string>()+splinesuffix);
+    sample_vecno.push_back(osc_channel["samplevecno"].as<int>());
+    sample_nutype.push_back(PDGToProbs(static_cast<NuPDG>(osc_channel["nutype"].as<int>())));
+    sample_oscnutype.push_back(PDGToProbs(static_cast<NuPDG>(osc_channel["oscnutype"].as<int>())));
+    sample_signal.push_back(osc_channel["signal"].as<bool>());
+  }
+
+  //Now get selection cuts
+  double low_bound = 0;
+  double up_bound = 0;
+  double KinematicParamter = 0;
+  std::vector<double> SelectionVec;
+  //Now grab the selection cuts from the manager
+  for ( auto const &SelectionCuts : SampleManager->raw()["SelectionCuts"]) {
+    SelectionStr.push_back(SelectionCuts["KinematicStr"].as<std::string>());
+    SelectionBounds.push_back(SelectionCuts["Bounds"].as<std::vector<double>>());
+    low_bound = SelectionBounds.back().at(0);
+    up_bound = SelectionBounds.back().at(1);
+    KinematicParamter = static_cast<double>(ReturnKinematicParameterFromString(SelectionCuts["KinematicStr"].as<std::string>()));
+    MACH3LOG_INFO("Adding cut on {} with bounds {} to {}", SelectionCuts["KinematicStr"].as<std::string>(), SelectionBounds.back().at(0), SelectionBounds.back().at(1));
+    SelectionVec = {KinematicParamter, low_bound, up_bound};
+    StoredSelection.push_back(SelectionVec);
+  }
+  NSelections = SelectionStr.size();
+
+  return;
+}
+
+void samplePDFFDBase::Initialise() {
```

Comment:
Yeah I don't like how clunky this is at the moment. You basically yo-yo between the base class and the derived class because the base class doesn't even know what to do into it's been told how much MC there is etc by the derived class. So it is quite backwards at the moment and stems from this slight confusion about the ownership of the underlying variables e.g. x_var is not actually owned by samplePDFFDBase it's owned by ,say, samplePDFDuneBeamFD. Probably this needs some restructuring where the simple reading of the MC is done by a separate object and then this gets passed to samplePDFDuneBeamFD and samplePDFFDBase by the constructor. I need to think about this more but I think it would clean up some of this bouncing about. I will raise this as an issue.

P.S. I don't like that this is called "Initiliasise" and that we also have an "Init" function...